lib/Dialect/Triton/Transforms/Combine.cpp

```diff
@@ -188,6 +188,43 @@ class CombineBroadcastMulReducePattern : public mlir::RewritePattern {
   }
 };
 
+// transpose(x, order=[0, 1, ...]) -> x
+class CombineNopTransPattern : public OpRewritePattern<TransOp> {
+public:
+  CombineNopTransPattern(mlir::MLIRContext *context)
+      : OpRewritePattern(context) {}
+
+  LogicalResult matchAndRewrite(TransOp trans,
+                                PatternRewriter &rewriter) const {
+    auto order = trans.getOrder();
+    for (int32_t i = 0; i < order.size(); i++) {
+      if (order[i] != i)
+        return failure();
+    }
+    rewriter.replaceAllUsesWith(trans, trans.getOperand());
+    return success();
+  }
+};
+
+// transpose(transpose(x)) -> transpose(x)
+class CombineNestedTransPattern : public OpRewritePattern<TransOp> {
+public:
+  CombineNestedTransPattern(MLIRContext *context) : OpRewritePattern(context) {}
+
+  mlir::LogicalResult matchAndRewrite(TransOp outer,
+                                      PatternRewriter &rewriter) const {
+    auto inner =
+        llvm::dyn_cast_or_null<TransOp>(outer.getOperand().getDefiningOp());
```

Comment:
more MLIR tricks nit :) can be
```
auto inner = outer.getOperand().getDefiningOp<TransOp>();
```

Reply:
That works even if the operand does not have a defining op?

Reply:
yes it would return a null reference

Reply:
Thanks, done.

Reply:
Oops, thanks :)